lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// A formatter and parser for [DateTime] in a fixed format [String] pattern.
+///
+/// For example, calling
+/// `FixedDateTimeCodec('YYYYMMDDhhmmss').decodeToLocal('19960425050322')` has
+/// the same result as calling `DateTime(1996, 4, 25, 5, 3, 22)`.
+///
+/// The allowed characters are
+/// * `Y`	a digit used in the time scale component “calendar year”
+/// * `M`	a digit used in the time scale component “calendar month”
+/// * `D`	a digit used in the time scale component “calendar day”
+/// * `E`	a digit used in the time scale component “decade”
+/// * `C`	a digit used in the time scale component “century”
+/// * `h`	a digit used in the time scale component “clock hour”
+/// * `m`	a digit used in the time scale component “clock minute”
+/// * `s`	a digit used in the time scale component “clock second”
+/// as specified in the ISO 8601 standard.
+///
+/// Non-allowed characters in the format [pattern] are ignored when decoding a
+/// string, in this case `YYYY kiwi MM` is the same format string as
+/// `YYYY------MM`. When encoding a datetime, the non-format characters are in
+/// the output verbatim.
+///
+/// Note: this class differs from [DateFormat] in that here, the characters are
+/// treated literally, i.e. the format string `YYY` matching `996` would result in
+/// the same as calling `DateTime(996)`. [DateFormat] on the other hand uses the
+/// specification in https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table,
+/// the format string (or "skeleton") `YYY` specifies only the padding, so
+/// `1996` would be a valid match. This limits it's use to format strings
+/// containing delimiters, as the parser would not know how many digits to take
+/// otherwise.
+class FixedDateTimeFormatter {
+  static final _validFormatCharacters = 'YMDEChms'.codeUnits;
+  static final yearCode = 'Y'.codeUnitAt(0);
```

Comment:
I'd make these constants. That allows the compiler to know the value, and you can use a switch below.

 static const yearCode = 0x59 /*Y*/;

(the codes are `0x59 /*Y*/`, `0x4D /*M*/`, `0x44 /*D*/`, `0x45 /*E*/`, `0x43 /*C*/`, `0x68 /*H*/`, `0x6D /*m*/`, `0x73 /*s*/`, and `0x53 /*S*/` if we want that - see below.)